include/aws/crt/Variant.h

```diff
@@ -0,0 +1,595 @@
+
+                using PlainT = typename std::decay<T>::type;
+                new (m_storage) PlainT(std::forward<Args>(args)...);
+                m_index = VariantDetail::Index::GetIndexOf<T, Ts...>();
+                AWS_ASSERT(m_index != -1);
+            }
+
+            Variant &operator=(const Variant &other)
+            {
+                if (this != &other)
+                {
+                    if (m_index != other.m_index)
+                    {
+                        Destroy();
+                        CopyConstructUtil<0, Ts...>(other);
+                    }
+                    else
+                    {
+                        CopyAssignUtil<0, Ts...>(other);
+                    }
+                }
+                return *this;
+            }
+
+            Variant &operator=(Variant &&other)
+            {
+                if (this != &other)
+                {
+                    if (m_index != other.m_index)
+                    {
+                        Destroy();
+                        MoveConstructUtil<0, Ts...>(std::move(other));
+                    }
+                    else
+                    {
+                        MoveAssignUtil<0, Ts...>(std::move(other));
+                    };
+                }
+                return *this;
+            }
+
+            /* emplace */
+            template <typename T, typename... Args, EnableIfOtherIsThisVariantAlternative<T> = 1>
+            T &emplace(Args &&...args)
+            {
+                static_assert(
+                    VariantDetail::Checker::HasType<typename std::decay<T>::type, Ts...>::value,
+                    "This variant does not have such alternative T.");
+                static_assert(
+                    sizeof(T) <= STORAGE_SIZE,
+                    "Attempting to instantiate a Variant with a type bigger than all alternatives.");
+
+                Destroy();
+
+                using PlainT = typename std::decay<T>::type;
+                new (m_storage) PlainT(std::forward<Args>(args)...);
+                m_index = VariantDetail::Index::GetIndexOf<T, Ts...>();
+                AWS_ASSERT(m_index != -1);
+
+                T *value = reinterpret_cast<T *>(m_storage);
+                return *value;
+            }
+
+            template <std::size_t Index, typename... Args>
+            auto emplace(Args &&...args) -> typename ThisVariantAlternative<Index>::type &
+            {
+                static_assert(Index < AlternativeCount, "Unknown alternative index to emplace");
+                using AlternativeT = typename ThisVariantAlternative<Index>::type;
+
+                return emplace<AlternativeT, Args...>(args...);
+            }
+
+            template <typename T, EnableIfOtherIsThisVariantAlternative<T> = 1> bool holds_alternative() const
+            {
+                AWS_ASSERT(m_index != -1);
+                return m_index == VariantDetail::Index::GetIndexOf<T, Ts...>();
```

Comment:
The [docs](https://en.cppreference.com/w/cpp/utility/variant/holds_alternative) say:
> The call is ill-formed if T does not appear exactly once in Types...

I genuinely don't know what "ill formed" means. Does that mean we should fail at compile time if T occurs twice in Types? Or does it mean that the implementation can do whatever it wants?

Reply:
ill formed means, [thanks to this stackoverflow reply](https://stackoverflow.com/a/15805532)
> If the Standard does not specify otherwise, an implementation should emit a diagnostic message (error or warning).

It is an undefined behavior. It is not very well specified how to handle repeated types in a variant. This variant implementation always visits left-to-right (except when directly accessed by Index, it is an UB in such case).